README
cheech - Chinese Checkers 
http://cheech.sourceforge.net/

cheech is a networked, cross-platform Chinese Checkers game supporting a few
rule variations (please tell me if you know of other fun rule variations!),
chatting with other players, and AI computer players.

cheech runs on on *nix systems, Windows, and MacOS, and there is a 
web-based client as well.

cheech is based on Loban Rahman's Gnome Chinese Checkers
(http://gchch.sourceforge.net/).


Using cheech
============

Basic cheech
------------
Run the program called cheech.  Choose the 'Host New Board' menu item and have
your friends connect to you.  Or use the 'Join a Board' menu item if someone 
else is already hosting a cheech board.

If your friends are busy, Host a New Board, and add computer players into the 
game using the 'Setup Computer Player...' and 'Add Computer Player' menu items.
Watch out, they're not super smart, but they are thorough!

To make a move, click on the peg you want to move, then on the empty hole to 
move your peg to.  If you're making a multiple-hop move, you need to click on 
each hole that you're hopping through along your chain of hops.  When you've
finished clicking out your hopping path, double click the last hole, or press 
Enter to make your move.


Advanced cheech
---------------
If you and your friends are all behind a port-blocking firewall, you can run
a separate cheech server on some publicly accessible server.  Run cheechd
on that server, and everyone can Join to that server's address.

You can also set up a cheechweb server so that people can join your game
just by browsing to a web site.  This requires a web server that can serve php
pages.
- make sure the people trying to connect to cheechweb can access the cheechweb
  port of your computer (port 3839 by default)
- start cheechd with the -W flag, or host a game from cheech.
  (Make sure to enable cheechweb from the Host a new Board dialog in cheech.)
- point people to the url of the cheechserver.  This will be:
  http://<computerHostingCheechweb>:<cheechwebPort>/
  (by default this is http://localhost:3839/ , where other people should replace
  localhost with the IP address of your computer.)


Building cheech
===============

Build on *nix/MacOS:
--------------
- install gtkmm (on a mac, run: brew install gtkmm)
- run the below commands in the root directory of the cheech project.
  - autoreconf -i
  - automake -a
  - ./configure
  - make
  - make install


Build on Windows using Dev-C++
------------------------------
- Get gtkmm 2.8 or higher from here:
    http://www.pcpm.ucl.ac.be/~gustin/win32_ports/gtkmm.html
  and install into C:\GTK.
- Use the included Dev-C++ files to build the cheech binaries.
- Put the cheech*.exe files and the pixmap directory into a folder together
- Rename the ajax/www/com.benjie.cheech directory cheechweb, and put it into
  the same folder as cheech.exe, and it should all work.


Ben Levitt 2005-2020


ChangeLog
2020-03-24  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.am (ACLOCAL_AMFLAGS): New variable.
	(EXTRA_DIST): Add m4/ChangeLog.


2006/2/7 9:51:52 PST
	update files from .glade file

2006/2/7 9:58:04 PST
	update files from .glade file

2006/2/7 10:04:47 PST
	update files from .glade file

2006/2/7 10:25:59 PST
	update files from .glade file

2006/2/7 10:47:38 PST
	update files from .glade file

2006/2/7 11:47:14 PST
	update files from .glade file

2006/2/7 11:50:09 PST
	update files from .glade file

2006/2/7 11:57:16 PST
	update files from .glade file

2006/2/7 12:02:36 PST
	update files from .glade file

2006/2/7 12:03:51 PST
	update files from .glade file

2006/2/7 12:06:19 PST
	update files from .glade file

2006/2/7 13:52:47 PST
	update files from .glade file

2006/2/7 13:58:44 PST
	update files from .glade file

2006/2/7 13:59:21 PST
	update files from .glade file

2006/2/7 14:01:13 PST
	update files from .glade file

2006/2/7 14:02:42 PST
	update files from .glade file

2006/2/8 14:18:49 PST
	update files from .glade file

2006/2/8 14:56:47 PST
	update files from .glade file

2006/2/8 16:39:49 PST
	update files from .glade file

2006/2/8 16:49:27 PST
	update files from .glade file

2006/2/9 19:37:17 PST
	update files from .glade file

2006/2/10 15:46:22 PST
	update files from .glade file

2006/2/10 15:47:29 PST
	update files from .glade file

2006/2/10 15:48:33 PST
	update files from .glade file

2006/2/10 16:16:07 PST
	update files from .glade file

2006/2/10 16:19:07 PST
	update files from .glade file

2006/2/10 17:41:59 PST
	update files from .glade file

2006/2/10 17:47:42 PST
	update files from .glade file

2006/2/14 12:30:13 PST
	update files from .glade file

2006/2/14 12:31:19 PST
	update files from .glade file

2006/2/14 12:46:30 PST
	update files from .glade file

2006/2/15 10:06:26 PST
	update files from .glade file

2006/2/15 10:18:13 PST
	update files from .glade file

2006/2/15 10:23:45 PST
	update files from .glade file

2006/2/15 10:24:46 PST
	update files from .glade file

2006/2/15 10:29:55 PST
	update files from .glade file

2006/2/15 10:42:35 PST
	update files from .glade file

2006/2/15 10:44:13 PST
	update files from .glade file

2006/2/15 12:23:33 PST
	update files from .glade file

2006/2/15 17:08:23 PST
	update files from .glade file

2006/2/15 17:30:56 PST
	update files from .glade file

2006/2/15 17:57:04 PST
	update files from .glade file

2006/2/15 18:05:53 PST
	update files from .glade file

2006/2/15 18:42:06 PST
	update files from .glade file

2006/2/15 18:54:21 PST
	update files from .glade file

2006/2/15 23:25:40 PST
	update files from .glade file

2006/2/16 23:23:58 PST
	update files from .glade file

2006/2/16 23:24:19 PST
	update files from .glade file

2006/2/16 23:48:05 PST
	update files from .glade file

2006/2/17 12:02:07 PST
	update files from .glade file

2006/2/17 12:03:21 PST
	update files from .glade file

2006/2/17 12:05:05 PST
	update files from .glade file

2006/2/18 14:26:21 PST
	update files from .glade file

2006/2/18 14:34:42 PST
	update files from .glade file

2006/2/18 17:01:15 PST
	update files from .glade file

2006/2/18 21:43:59 PST
	update files from .glade file

2006/2/18 23:42:43 PST
	update files from .glade file

2006/2/28 18:26:31 PST
	update files from .glade file

2006/2/28 18:28:50 PST
	update files from .glade file

2006/3/8 12:52:39 PST
	update files from .glade file

2006/3/8 12:59:04 PST
	update files from .glade file

2006/3/12 22:31:54 PST
	update files from .glade file

2006/3/12 23:40:02 PST
	update files from .glade file

2006/3/12 23:50:30 PST
	update files from .glade file

2006/3/13 0:16:53 PST
	update files from .glade file

2006/3/13 0:20:53 PST
	update files from .glade file

2006/3/13 14:44:53 PST
	update files from .glade file

2006/3/13 14:54:18 PST
	update files from .glade file

2006/3/22 10:51:39 PST
	update files from .glade file

2006/3/22 11:01:30 PST
	update files from .glade file

2006/3/23 21:38:28 PST
	update files from .glade file

2006/3/23 21:57:20 PST
	update files from .glade file

2006/3/23 22:01:04 PST
	update files from .glade file

2006/3/23 22:21:44 PST
	update files from .glade file

2006/3/23 22:40:59 PST
	update files from .glade file

2006/3/24 23:42:22 PST
	update files from .glade file

2006/3/25 0:35:56 PST
	update files from .glade file

2006/3/26 14:57:54 PST
	update files from .glade file

2006/3/26 15:15:54 PST
	update files from .glade file

2006/3/26 15:21:41 PST
	update files from .glade file

2006/3/26 15:22:11 PST
	update files from .glade file

2006/3/26 15:33:00 PST
	update files from .glade file

2006/3/26 15:41:53 PST
	update files from .glade file

2006/3/26 15:45:11 PST
	update files from .glade file

2006/3/26 16:24:41 PST
	update files from .glade file

2006/3/26 16:27:26 PST
	update files from .glade file

2006/3/26 16:38:36 PST
	update files from .glade file

2006/3/26 16:43:39 PST
	update files from .glade file

2006/3/26 16:57:52 PST
	update files from .glade file

2006/3/26 17:49:40 PST
	update files from .glade file

2006/3/28 12:58:24 PST
	update files from .glade file

2006/3/28 13:05:16 PST
	update files from .glade file

2006/3/28 14:09:28 PST
	update files from .glade file

2006/3/28 14:12:48 PST
	update files from .glade file

2006/3/28 21:07:42 PST
	update files from .glade file

2006/3/28 21:12:46 PST
	update files from .glade file

2006/3/29 10:32:41 PST
	update files from .glade file

2006/4/4 19:20:49 EST
	update files from .glade file

2006/4/4 20:44:14 EST
	update files from .glade file

2006/4/4 20:45:07 EST
	update files from .glade file

2006/4/5 0:19:14 EST
	update files from .glade file

2006/4/5 23:13:24 EST
	update files from .glade file

2006/4/5 23:16:02 EST
	update files from .glade file

2006/4/6 19:33:50 EST
	update files from .glade file

2006/4/6 19:56:32 EST
	update files from .glade file

2006/4/7 0:12:25 EST
	update files from .glade file

2006/4/7 0:19:22 EST
	update files from .glade file


src/cheechd.cc
/*
 *  cheechd application's main.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 */

#include <iostream>
#include <string>
#include <stdlib.h>
#include <config.h>
#include <glib/gi18n.h>
#include <glibmm/main.h>
#include <glibmm/iochannel.h>
#include <glibmm/optioncontext.h>
#include <sigc++/bind.h>
#include <gnet-2.0/gnet.h>

#include "utility.hh"
#include "game_server.hh"
#include "ajax_server.hh"


// cheechd Options
int num_players;
int num_games;
int port;
bool long_jumps;
bool hop_others;
bool stop_others;

bool start_cheechwebd;
int cheechweb_port;


void printMessage(Glib::ustring msg)
{
	std::cout << msg << std::endl;
}


void gameOver(GameServer *server, Glib::RefPtr<Glib::MainLoop> m)
{
	num_games--;

	if (num_games)
		server->rotate_players();
	else
		m->quit();
}


bool on_stdin(Glib::IOCondition condition, GameServer *server)
{
	std::string line;

	if(!std::getline(std::cin, line))
	{
		server->end_game();
		exit(0);
	}

	server->parse_command(line);
	return true;
}


void process_options(int &argc, char **&argv)
{
	try
	{
		Glib::OptionContext opt_context;
		Glib::OptionGroup opt_group(
			"cheechd server options", "Options defining the game being served");

		Glib::OptionEntry opt_port;
		opt_port.set_long_name("port");
		opt_port.set_short_name('p');
		opt_port.set_arg_description("port");
		opt_port.set_description(
			"port to use for the cheech server (3838)");
		opt_group.add_entry(opt_port, port);

		Glib::OptionEntry opt_num_players;
		opt_num_players.set_long_name("num-players");
		opt_num_players.set_short_name('N');
		opt_num_players.set_arg_description("num");
		opt_num_players.set_description(
			"how many players will be in this game (3)");
		opt_group.add_entry(opt_num_players, num_players);

		Glib::OptionEntry opt_num_games;
		opt_num_games.set_long_name("num-games");
		opt_num_games.set_short_name('g');
		opt_num_games.set_arg_description("num");
		opt_num_games.set_description(
			"play this many games and then stop");
		opt_group.add_entry(opt_num_games, num_games);

		Glib::OptionEntry opt_long_jumps;
		opt_long_jumps.set_long_name("long-jumps");
		opt_long_jumps.set_short_name('L');
		opt_long_jumps.set_description(
			"allow long jumps (default=no)");
		opt_group.add_entry(opt_long_jumps, long_jumps);

		Glib::OptionEntry opt_hop_others;
		opt_hop_others.set_long_name("hop-others");
		opt_hop_others.set_short_name('H');
		opt_hop_others.set_description(
			"allow hopping through other players' triangles (default=no)");
		opt_group.add_entry(opt_hop_others, hop_others);

		Glib::OptionEntry opt_stop_others;
		opt_stop_others.set_long_name("stop-others");
		opt_stop_others.set_short_name('O');
		opt_stop_others.set_description(
			"allow stopping in other players' triangles (default=no)");
		opt_group.add_entry(opt_stop_others, stop_others);

		Glib::OptionEntry opt_start_cheechwebd;
		opt_start_cheechwebd.set_long_name("start-cheechweb");
		opt_start_cheechwebd.set_short_name('W');
		opt_start_cheechwebd.set_description(
			"start the cheechweb server (default=no)");
		opt_group.add_entry(opt_start_cheechwebd, start_cheechwebd);

		Glib::OptionEntry opt_cheechweb_port;
		opt_cheechweb_port.set_long_name("cheechweb-port");
		opt_cheechweb_port.set_short_name('P');
		opt_cheechweb_port.set_arg_description("port");
		opt_cheechweb_port.set_description(
			"port to use for the cheechweb server (3839)");
		opt_group.add_entry(opt_cheechweb_port, cheechweb_port);

		opt_context.set_main_group(opt_group);

		opt_context.parse(argc, argv);
	}
	catch (Glib::OptionError er)
	{
		std::cout << "Bad command line arguments.  Try cheechd --help"
			<< std::endl;
		exit(1);
	}

	// Defaults
	if (port == 0) port = 3838;
	if (cheechweb_port == 0 && start_cheechwebd) cheechweb_port = 3839;
	if (num_players == 0) num_players = 3;
	if (num_players < 1) num_players = 1;
	if (num_players > 6) num_players = 6;
}


int main(int argc, char **argv)
{
#if defined(ENABLE_NLS)
	bindtextdomain (GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);
#endif //ENABLE_NLS

	gnet_init();

	process_options(argc, argv);

	Glib::RefPtr<Glib::MainLoop> m = Glib::MainLoop::create();

	GameServer *server = new GameServer(port, num_players, long_jumps,
										hop_others, stop_others);
	AjaxServer *ajax_server;

	server->evt_message.connect(sigc::ptr_fun(printMessage));
	if (num_games)
		server->evt_game_over.connect(sigc::bind(sigc::bind(sigc::ptr_fun(gameOver), m), server));
	server->new_game();

	if (!server->ready())
	{
		delete server;
		std::cout << "Chinese Checkers GameServer failed to start on port "
			<< port << std::endl;
		return 0;
	}

	std::cout << "Chinese Checkers GameServer started on port "
		<< port << std::endl;

	if (cheechweb_port)
	{
		ajax_server = new AjaxServer();
		if (!ajax_server->listen(cheechweb_port, "localhost", port))
		{
			delete ajax_server;
			std::cout << "CheechWeb Server failed to start on port "
				<< cheechweb_port << std::endl;
			ajax_server = NULL;
		}
		else
		{
			std::cout << "CheechWeb Server started on port "
				<< cheechweb_port << std::endl;
		}
	}


	std::cout << "Game started for " << num_players << " players." << std::endl;
	std::cout << "Long Jumps" << (long_jumps?" ":" not ") << "allowed."
		<< std::endl;
	std::cout << "Hopping through other players' triangles"
		<< (hop_others?" ":" not ") << "allowed." << std::endl;
	std::cout << "Stopping in other players' triangles"
		<< (stop_others?" ":" not ") << "allowed." << std::endl;

	std::cout << std::endl;

	Glib::signal_io().connect(sigc::bind(sigc::ptr_fun(on_stdin), server),
							  Glib::IOChannel::create_from_fd(0),
							  Glib::IO_IN);

	m->run();

	return 0;
}
